File: src/purra/artifacts/lifecycle.py
"""Core-owned state machine for recoverable artifact commits."""

from __future__ import annotations

import uuid
from collections.abc import Callable, Sequence

from purra.artifacts.contracts import (
    ArtifactAppendCommand,
    ArtifactBatch,
    ArtifactBatchReceipt,
    ArtifactCreateCommand,
    ArtifactFinalizeCommand,
    ArtifactMutationLease,
    ArtifactRecord,
    ArtifactStatus,
    ArtifactValidationResult,
    coverage_digest,
)
from purra.artifacts.errors import (
    ArtifactConflictError,
    ArtifactNotFoundError,
    ArtifactStateError,
    ArtifactValidationError,
)
from purra.artifacts.ports import ArtifactRepository, ArtifactValidator


class ArtifactLifecycle:
    """Coordinate validation while repositories guarantee atomic CAS writes."""

    def __init__(
        self,
        repository: ArtifactRepository,
        *,
        validator: ArtifactValidator | None = None,
        id_factory: Callable[[], str] | None = None,
    ) -> None:
        self._repository = repository
        self._validator = validator
        self._id_factory = id_factory or (lambda: uuid.uuid4().hex)

    async def begin(self, command: ArtifactCreateCommand) -> ArtifactRecord:
        artifact_id = str(self._id_factory() or "").strip()
        if not artifact_id:
            raise RuntimeError("artifact id factory returned an empty id")
        return await self._repository.create(artifact_id, command)

    async def get(self, artifact_id: str) -> ArtifactRecord:
        """Load an artifact or raise the same typed error as mutations."""

        return await self._require(artifact_id)

    async def append(
        self,
        command: ArtifactAppendCommand,
    ) -> ArtifactBatchReceipt:
        replayed = await self._repository.replay_receipt(command)
        if replayed is not None:
            return replayed
        artifact = await self._require(command.artifact_id)
        _require_open(artifact)
        if artifact.revision != command.expected_revision:
            raise ArtifactConflictError(
                "artifact revision does not match",
                code="artifact_revision_conflict",
                details={
                    "expectedRevision": command.expected_revision,
                    "actualRevision": artifact.revision,
                },
            )
        if artifact.next_sequence != command.sequence:
            raise ArtifactConflictError(
                "artifact batch sequence is not next",
                code="artifact_sequence_conflict",
                details={
                    "expectedSequence": artifact.next_sequence,
                    "actualSequence": command.sequence,
                },
            )
        duplicate_coverage = _duplicates(command.coverage_keys)
        if duplicate_coverage:
            raise ArtifactValidationError(
                "artifact batch contains duplicate coverage keys",
                code="artifact_coverage_duplicate",
                details={"duplicateKeys": sorted(duplicate_coverage)},
            )
        committed_coverage = {
            key
            for batch in await self._repository.list_batches(artifact.id)
            for key in batch.coverage_keys
        }
        repeated_coverage = committed_coverage.intersection(
            command.coverage_keys
        )
        if repeated_coverage:
            raise ArtifactValidationError(
                "artifact batch repeats committed coverage keys",
                code="artifact_coverage_duplicate",
                details={"duplicateKeys": sorted(repeated_coverage)},
            )
        await self._validate_batch(artifact, command)
        return await self._repository.append(command)

    async def finalize(
        self,
        command: ArtifactFinalizeCommand,
    ) -> ArtifactRecord:
        artifact = await self._require(command.artifact_id)
        _require_open(artifact)
        if artifact.revision != command.expected_revision:
            raise ArtifactConflictError(
                "artifact revision does not match",
                code="artifact_revision_conflict",
                details={
                    "expectedRevision": command.expected_revision,
                    "actualRevision": artifact.revision,
                },
            )
        expected_count = (
            command.expected_item_count
            if command.expected_item_count is not None
            else artifact.expected_item_count
        )
        if (
            artifact.expected_item_count is not None
            and command.expected_item_count is not None
            and artifact.expected_item_count != command.expected_item_count
        ):
            raise ArtifactValidationError(
                "final item count conflicts with the artifact manifest",
                code="artifact_manifest_count_mismatch",
                details={
                    "manifestCount": artifact.expected_item_count,
                    "finalizeCount": command.expected_item_count,
                },
            )
        if (
            expected_count is not None
            and artifact.committed_item_count != expected_count
        ):
            raise ArtifactValidationError(
                "artifact item count is incomplete",
                code="artifact_item_count_incomplete",
                details={
                    "expectedCount": expected_count,
                    "actualCount": artifact.committed_item_count,
                },
            )
        batches = tuple(await self._repository.list_batches(artifact.id))
        _validate_batch_sequence(batches)
        observed_coverage = tuple(
            key
            for batch in batches
            for key in batch.coverage_keys
        )
        duplicates = _duplicates(observed_coverage)
        if duplicates:
            raise ArtifactValidationError(
                "artifact coverage contains duplicate keys",
                code="artifact_coverage_duplicate",
                details={"duplicateKeys": sorted(duplicates)},
            )
        expected_coverage = set(command.expected_coverage_keys)
        observed_coverage_set = set(observed_coverage)
        if expected_coverage and observed_coverage_set != expected_coverage:
            raise ArtifactValidationError(
                "artifact coverage does not match the final manifest",
                code="artifact_coverage_mismatch",
                details={
                    "missingKeys": sorted(expected_coverage - observed_coverage_set),
                    "unexpectedKeys": sorted(
                        observed_coverage_set - expected_coverage
                    ),
                },
            )
        await self._validate_finalization(artifact, batches, command)
        return await self._repository.finalize(
            command,
            coverage_digest=coverage_digest(tuple(observed_coverage)),
        )

    async def abort(
        self,
        artifact_id: str,
        *,
        expected_revision: int,
        write_lease: ArtifactMutationLease,
    ) -> ArtifactRecord:
        if not isinstance(write_lease, ArtifactMutationLease):
            raise TypeError("artifact abort write_lease is invalid")
        artifact = await self._require(artifact_id)
        _require_open(artifact)
        if artifact.revision != int(expected_revision):
            raise ArtifactConflictError(
                "artifact revision does not match",
                code="artifact_revision_conflict",
                details={
                    "expectedRevision": int(expected_revision),
                    "actualRevision": artifact.revision,
                },
            )
        return await self._repository.abort(
            artifact.id,
            expected_revision=artifact.revision,
            write_lease=write_lease,
        )

    async def _require(self, artifact_id: str) -> ArtifactRecord:
        normalized = str(artifact_id or "").strip()
        artifact = await self._repository.load(normalized)
        if artifact is None:
            raise ArtifactNotFoundError(
                "artifact does not exist",
                code="artifact_not_found",
                details={"artifactId": normalized},
            )
        return artifact

    async def _validate_batch(
        self,
        artifact: ArtifactRecord,
        command: ArtifactAppendCommand,
    ) -> None:
        if self._validator is None:
            return
        result = await self._validator.validate_batch(artifact, command)
        _require_validation_result(result)

    async def _validate_finalization(
        self,
        artifact: ArtifactRecord,
        batches: Sequence[ArtifactBatch],
        command: ArtifactFinalizeCommand,
    ) -> None:
        if self._validator is None:
            return
        result = await self._validator.validate_finalization(
            artifact,
            batches,
            command,
        )
        _require_validation_result(result)


def _require_open(artifact: ArtifactRecord) -> None:
    if artifact.status is not ArtifactStatus.OPEN:
        raise ArtifactStateError(
            "artifact is not open",
            code="artifact_not_open",
            details={"status": artifact.status.value},
        )


def _require_validation_result(result: ArtifactValidationResult) -> None:
    if not isinstance(result, ArtifactValidationResult):
        raise TypeError("artifact validator returned an invalid result")
    if not result.accepted:
        raise ArtifactValidationError(
            "artifact domain validation rejected the operation",
            code=result.code or "artifact_validation_failed",
            details=result.details,
        )


def _validate_batch_sequence(batches: Sequence[ArtifactBatch]) -> None:
    observed = tuple(batch.sequence for batch in batches)
    expected = tuple(range(1, len(batches) + 1))
    if observed != expected:
        raise ArtifactValidationError(
            "artifact batch sequence contains a gap or reordering",
            code="artifact_batch_sequence_invalid",
            details={"expected": expected, "actual": observed},
        )


def _duplicates(values: Sequence[str]) -> set[str]:
    seen: set[str] = set()
    duplicates: set[str] = set()
    for value in values:
        if value in seen:
            duplicates.add(value)
        seen.add(value)
    return duplicates


__all__ = ["ArtifactLifecycle"]

File: typescript/src/artifacts/lifecycle.ts
import { AgentError } from "../shared/errors.js";
import {
  artifactCoverageDigest,
  copyArtifactCreateCommand,
  copyArtifactFinalizeCommand,
  copyArtifactMutationLease,
  copyArtifactValidationResult,
  positiveInteger,
  prepareArtifactAppend,
  requiredText,
} from "./contracts.js";
import type {
  ArtifactAppendCommand,
  ArtifactBatch,
  ArtifactBatchReceipt,
  ArtifactCreateCommand,
  ArtifactFinalizeCommand,
  ArtifactRecord,
  ArtifactRepository,
  ArtifactValidator,
} from "./types.js";

export class ArtifactLifecycle {
  readonly #repository: ArtifactRepository;
  readonly #validator: ArtifactValidator | undefined;
  readonly #idFactory: () => string;

  public constructor(options: {
    readonly repository: ArtifactRepository;
    readonly validator?: ArtifactValidator;
    readonly idFactory?: () => string;
  }) {
    if (
      typeof options.repository?.create !== "function"
      || typeof options.repository.append !== "function"
      || typeof options.repository.finalize !== "function"
    ) {
      throw new TypeError("Artifact lifecycle requires a repository");
    }
    if (
      options.validator !== undefined
      && (
        typeof options.validator.validateBatch !== "function"
        || typeof options.validator.validateFinalization !== "function"
      )
    ) {
      throw new TypeError("Artifact validator is incomplete");
    }
    this.#repository = options.repository;
    this.#validator = options.validator;
    this.#idFactory = options.idFactory ?? (() => globalThis.crypto.randomUUID());
  }

  public async begin(raw: ArtifactCreateCommand): Promise<ArtifactRecord> {
    const command = copyArtifactCreateCommand(raw);
    const artifactId = requiredText(this.#idFactory(), "Artifact id");
    return this.#repository.create(artifactId, command);
  }

  public get(artifactId: string): Promise<ArtifactRecord> {
    return this.#require(artifactId);
  }

  public async append(raw: ArtifactAppendCommand): Promise<ArtifactBatchReceipt> {
    const command = await prepareArtifactAppend(raw);
    const replayed = await this.#repository.replayReceipt(command);
    if (replayed !== undefined) return replayed;
    const artifact = await this.#require(command.artifactId);
    requireOpen(artifact);
    requireRevision(artifact, command.expectedRevision);
    if (artifact.nextSequence !== command.sequence) {
      throw new AgentError("artifact_sequence_conflict", "Artifact batch is not next");
    }
    const duplicateInBatch = duplicates(command.coverageKeys);
    if (duplicateInBatch.length > 0) {
      throw new AgentError("artifact_coverage_duplicate", "Artifact batch repeats coverage keys");
    }
    const batches = await this.#repository.listBatches(artifact.id);
    const committedCoverage = new Set(batches.flatMap((batch) => batch.coverageKeys));
    if (command.coverageKeys.some((key) => committedCoverage.has(key))) {
      throw new AgentError("artifact_coverage_duplicate", "Artifact coverage was already committed");
    }
    await this.#validateBatch(artifact, command);
    return this.#repository.append(command);
  }

  public async finalize(raw: ArtifactFinalizeCommand): Promise<ArtifactRecord> {
    const command = copyArtifactFinalizeCommand(raw);
    const artifact = await this.#require(command.artifactId);
    requireOpen(artifact);
    requireRevision(artifact, command.expectedRevision);
    if (
      artifact.expectedItemCount !== null
      && command.expectedItemCount !== null
      && artifact.expectedItemCount !== command.expectedItemCount
    ) {
      throw new AgentError(
        "artifact_manifest_count_mismatch",
        "Artifact final item count conflicts with its manifest",
      );
    }
    const expectedCount = command.expectedItemCount ?? artifact.expectedItemCount;
    if (expectedCount !== null && artifact.committedItemCount !== expectedCount) {
      throw new AgentError("artifact_item_count_incomplete", "Artifact item count is incomplete");
    }
    const batches = await this.#repository.listBatches(artifact.id);
    validateSequence(batches);
    const coverage = batches.flatMap((batch) => batch.coverageKeys);
    if (duplicates(coverage).length > 0) {
      throw new AgentError("artifact_coverage_duplicate", "Artifact coverage contains duplicates");
    }
    if (command.expectedCoverageKeys.length > 0) {
      const expected = new Set(command.expectedCoverageKeys);
      const observed = new Set(coverage);
      if (
        expected.size !== observed.size
        || [...expected].some((key) => !observed.has(key))
      ) {
        throw new AgentError("artifact_coverage_mismatch", "Artifact coverage is incomplete");
      }
    }
    await this.#validateFinalization(artifact, batches, command);
    return this.#repository.finalize(command, await artifactCoverageDigest(coverage));
  }

  public async abort(input: {
    readonly artifactId: string;
    readonly expectedRevision: number;
    readonly writeLease: import("./types.js").ArtifactMutationLease;
  }): Promise<ArtifactRecord> {
    const artifactId = requiredText(input.artifactId, "Artifact id");
    const expectedRevision = positiveInteger(input.expectedRevision, "Artifact expected revision");
    const writeLease = copyArtifactMutationLease(input.writeLease);
    const artifact = await this.#require(artifactId);
    requireOpen(artifact);
    requireRevision(artifact, expectedRevision);
    return this.#repository.abort({ artifactId, expectedRevision, writeLease });
  }

  async #require(artifactId: string): Promise<ArtifactRecord> {
    const normalized = requiredText(artifactId, "Artifact id");
    const artifact = await this.#repository.load(normalized);
    if (artifact === undefined) throw new AgentError("artifact_not_found", "Artifact does not exist");
    return artifact;
  }

  async #validateBatch(
    artifact: ArtifactRecord,
    command: import("./types.js").ArtifactAppendOperation,
  ): Promise<void> {
    if (this.#validator === undefined) return;
    const result = copyArtifactValidationResult(
      await this.#validator.validateBatch(artifact, command),
    );
    if (!result.accepted) {
      throw new AgentError(result.code, "Artifact batch validation failed");
    }
  }

  async #validateFinalization(
    artifact: ArtifactRecord,
    batches: readonly ArtifactBatch[],
    command: import("./types.js").ArtifactFinalizeOperation,
  ): Promise<void> {
    if (this.#validator === undefined) return;
    const result = copyArtifactValidationResult(
      await this.#validator.validateFinalization(artifact, batches, command),
    );
    if (!result.accepted) {
      throw new AgentError(result.code, "Artifact finalization validation failed");
    }
  }
}

function requireOpen(artifact: ArtifactRecord): void {
  if (artifact.status !== "open") {
    throw new AgentError("artifact_not_open", "Artifact is not open");
  }
}

function requireRevision(artifact: ArtifactRecord, expected: number): void {
  if (artifact.revision !== expected) {
    throw new AgentError("artifact_revision_conflict", "Artifact revision conflicts");
  }
}

function validateSequence(batches: readonly ArtifactBatch[]): void {
  if (batches.some((batch, index) => batch.sequence !== index + 1)) {
    throw new AgentError("artifact_batch_sequence_invalid", "Artifact batch sequence is invalid");
  }
}

function duplicates(values: readonly string[]): readonly string[] {
  const seen = new Set<string>();
  const repeated = new Set<string>();
  for (const value of values) {
    if (seen.has(value)) repeated.add(value);
    seen.add(value);
  }
  return Object.freeze([...repeated].sort());
}
